src/loader/module_wrap.cc

```diff
@@ -0,0 +1,492 @@
+  TryCatch evalCatch(iso);
+  auto result = obj->module_.Get(iso)->Evaluate(ctx);
+  if (evalCatch.HasCaught()) {
+    evalCatch.ReThrow();
+    return;
+  }
+  auto ret = result.ToLocalChecked();
+  args.GetReturnValue().Set(ret);
+}
+
+MaybeLocal<Module> ModuleWrap::ResolveCallback(Local<Context> context,
+                                                       Local<String> specifier,
+                                                       Local<Module> referrer) {
+  Environment* env = Environment::GetCurrent(context);
+  Isolate* iso = Isolate::GetCurrent();
+  if (ModuleWrap::module_map_.count(referrer->GetIdentityHash()) == 0) {
+    env->ThrowError("linking error, unknown module");
+    return MaybeLocal<Module>();
+  }
+
+  auto possible_deps = ModuleWrap::module_map_[referrer->GetIdentityHash()];
+  ModuleWrap* dependent = nullptr;
+
+  for (auto possible_dep : *possible_deps) {
+    if (possible_dep->module_ == referrer) {
+      dependent = possible_dep;
+    }
+  }
+
+  String::Utf8Value ss(specifier);
+
+  if (dependent == nullptr) {
+    env->ThrowError("linking error, null dep");
+    return MaybeLocal<Module>();
+  }
+
+  String::Utf8Value utf8_specifier(specifier);
+  std::string std_specifier = *utf8_specifier;
+
+  if (dependent->resolve_cache_.count(std_specifier) != 1) {
+    env->ThrowError("linking error, not in local cache");
+    return MaybeLocal<Module>();
+  }
+
+  Local<Promise> resolvePromise = dependent->resolve_cache_[std_specifier]->Get(iso);
+
+  if (resolvePromise->State() != Promise::kFulfilled) {
+    env->ThrowError("linking error, dependency promises must be resolved on instantiate");
+    return MaybeLocal<Module>();
+  }
+
+  MaybeLocal<Object> moduleObject = resolvePromise->Result()->ToObject();
+
+  if (moduleObject.IsEmpty()) {
+    env->ThrowError("linking error, expected a valid module object from resolver");
+    return MaybeLocal<Module>();
+  }
+
+  ModuleWrap* mod = Unwrap<ModuleWrap>(moduleObject.ToLocalChecked());
+
+  return mod->module_.Get(context->GetIsolate());
+}
+
+using namespace node::url;
+  URL __init_cwd() {
+    std::string specifier = "file://";
+    #ifdef _WIN32
+      /* MAX_PATH is in characters, not bytes. Make sure we have enough headroom. */
+      char buf[MAX_PATH * 4];
+    #else
+      char buf[PATH_MAX];
+    #endif
+
+    size_t cwd_len = sizeof(buf);
+    int err = uv_cwd(buf, &cwd_len);
+    if (err) {
+      return URL("");
+    }
+    specifier += buf;
+    specifier += "/";
+    return URL(specifier);
+  }
+  static node::url::URL INITIAL_CWD(__init_cwd());
+  inline bool is_relative_or_absolute_path(std::string specifier) {
+    auto len = specifier.length();
+    if (len <= 0) return false;
+    else if (specifier[0] == '/') return true;
+    else if (specifier[0] == '.') {
+      if (len == 1 || specifier[1] == '/') return true;
+      else if (specifier[1] == '.') {
+        if (len == 2 || specifier[2] == '/') return true;
```

Comment:
Never mind, this is to be consistent with https://html.spec.whatwg.org/multipage/webappapis.html#resolve-a-module-specifier.